File: mm-rs/src/mm/decoder.rs
use crate::jsonc::ast::{Node, Object, Array, Value, Field, ValueData};
use crate::jsonc::tag::Tag;
use crate::jsonc::ValueType;
use crate::mm::prefix::Prefix;
use crate::mm::simple_value::SimpleValue;

pub struct Decoder {
    data: Vec<u8>,
    offset: usize,
}

impl Decoder {
    pub fn new(data: Vec<u8>) -> Self {
        Self { data, offset: 0 }
    }

    fn read_byte(&mut self) -> Result<u8, std::io::Error> {
        if self.offset >= self.data.len() {
            return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "unexpected eof"));
        }
        let b = self.data[self.offset];
        self.offset += 1;
        Ok(b)
    }

    fn read_bytes(&mut self, n: usize) -> Result<&[u8], std::io::Error> {
        if self.offset + n > self.data.len() {
            return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "unexpected eof"));
        }
        let start = self.offset;
        self.offset += n;
        Ok(&self.data[start..start + n])
    }

    pub fn decode(&mut self) -> Result<Node, std::io::Error> {
        let tag = Tag::new();
        self.decode_node(&tag)
    }

    fn decode_node(&mut self, tag: &Tag) -> Result<Node, std::io::Error> {
        let b = self.read_byte()?;
        let prefix = Prefix::from_byte(b).ok_or_else(|| {
            std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid prefix")
        })?;

        match prefix {
            Prefix::Tag => self.decode_tag(b),
            Prefix::Simple => self.decode_simple(b, tag),
            Prefix::PositiveInt => self.decode_positive_int(b, tag),
            Prefix::NegativeInt => self.decode_negative_int(b, tag),
            Prefix::PrefixFloat => self.decode_float(b, tag),
            Prefix::PrefixString => self.decode_string(b, tag),
            Prefix::PrefixBytes => self.decode_bytes(b, tag),
            Prefix::Container => self.decode_container(b, tag),
        }
    }

    fn decode_tag(&mut self, prefix: u8) -> Result<Node, std::io::Error> {
        let l1 = match prefix & 0x03 {
            0 => 0,
            1 => {
                let l = self.read_byte()?;
                l as usize
            }
            2 => {
                let l = self.read_bytes(2)?;
                ((l[0] as usize) << 8) | (l[1] as usize)
            }
            _ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid tag length")),
        };

        let mut tag = Tag::new();

        let b = self.read_byte()?;
        let mut l = b as usize;

        if l < 254 {
        } else if l < 257 {
            l = self.read_byte()? as usize;
        } else {
            let l3 = self.read_bytes(2)?;
            l = ((l3[0] as usize) << 8) | (l3[1] as usize);
        }

        for _ in 0..l {
            self.decode_tag_bytes(&mut tag)?;
        }

        if tag.is_null {
            let data = match tag.value_type {
                ValueType::Int => ValueData::Int(0),
                ValueType::Float64 => ValueData::Float(0.0),
                ValueType::Bool => ValueData::Bool(false),
                ValueType::String => ValueData::String(String::new()),
                ValueType::Bytes => ValueData::Bytes(vec![]),
                ValueType::DateTime => ValueData::String("1970-01-01T00:00:00Z".to_string()),
                _ => ValueData::Null,
            };
            Ok(Node::Value(Value {
                data,
                text: String::new(),
                tag: Some(tag),
            }))
        } else {
            self.decode_node(&tag)
        }
    }

    fn decode_tag_bytes(&mut self, _tag: &mut Tag) -> Result<usize, std::io::Error> {
        Ok(0)
    }

    fn decode_simple(&mut self, prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let value = SimpleValue::from_byte(prefix & 0x1F).ok_or_else(|| {
            std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid simple value")
        })?;

        let (data, text) = match value {
            SimpleValue::False => (ValueData::Bool(false), "false".to_string()),
            SimpleValue::True => (ValueData::Bool(true), "true".to_string()),
            SimpleValue::NullBool => (ValueData::Bool(false), "false".to_string()),
            SimpleValue::NullFloat => (ValueData::Float(0.0), "0.0".to_string()),
            SimpleValue::NullInt => (ValueData::Int(0), "0".to_string()),
            SimpleValue::NullString => (ValueData::String(String::new()), String::new()),
            SimpleValue::NullBytes => (ValueData::Bytes(vec![]), String::new()),
            _ => (ValueData::String(String::new()), String::new()),
        };

        Ok(Node::Value(Value {
            data,
            text,
            tag: Some(Tag::new()),
        }))
    }

    fn decode_positive_int(&mut self, prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let l1 = match prefix & 0x07 {
            0 => 0,
            1 => {
                let l = self.read_byte()?;
                l as usize
            }
            _ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid int length")),
        };

        let mut v = 0u64;
        for i in 0..l1 {
            let b = self.read_byte()?;
            v |= (b as u64) << (i * 8);
        }

        let data = ValueData::Int(v as i64);
        let text = v.to_string();

        Ok(Node::Value(Value {
            data,
            text,
            tag: Some(Tag::new()),
        }))
    }

    fn decode_negative_int(&mut self, prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let l1 = match prefix & 0x07 {
            0 => 0,
            1 => {
                let l = self.read_byte()?;
                l as usize
            }
            _ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid int length")),
        };

        let mut v = 0u64;
        for i in 0..l1 {
            let b = self.read_byte()?;
            v |= (b as u64) << (i * 8);
        }

        let data = ValueData::Int(-(v as i64));
        let text = format!("-{}", v);

        Ok(Node::Value(Value {
            data,
            text,
            tag: Some(Tag::new()),
        }))
    }

    fn decode_float(&mut self, _prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let s = String::from_utf8_lossy(self.read_bytes(8)?).to_string();
        let f = s.parse::<f64>().unwrap_or(0.0);
        Ok(Node::Value(Value {
            data: ValueData::Float(f),
            text: ryu::Buffer::new().format_finite(f).to_string(),
            tag: Some(Tag::new()),
        }))
    }

    fn decode_string(&mut self, prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let l1 = match prefix & 0x03 {
            0 => 0,
            1 => {
                let l = self.read_byte()?;
                l as usize
            }
            2 => {
                let l = self.read_bytes(2)?;
                ((l[0] as usize) << 8) | (l[1] as usize)
            }
            _ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid string length")),
        };

        let s = if l1 > 0 {
            String::from_utf8_lossy(self.read_bytes(l1)?).to_string()
        } else {
            String::new()
        };

        Ok(Node::Value(Value {
            data: ValueData::String(s.clone()),
            text: s,
            tag: Some(Tag::new()),
        }))
    }

    fn decode_bytes(&mut self, prefix: u8, _tag: &Tag) -> Result<Node, std::io::Error> {
        let l1 = match prefix & 0x03 {
            0 => 0,
            1 => {
                let l = self.read_byte()?;
                l as usize
            }
            2 => {
                let l = self.read_bytes(2)?;
                ((l[0] as usize) << 8) | (l[1] as usize)
            }
            _ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid bytes length")),
        };

        let bytes = if l1 > 0 {
            self.read_bytes(l1)?.to_vec()
        } else {
            vec![]
        };

        Ok(Node::Value(Value {
            data: ValueData::Bytes(bytes.clone()),
            text: format!("{:?}", bytes),
            tag: Some(Tag::new()),
        }))
    }

    fn decode_container(&mut self, prefix: u8, tag: &Tag) -> Result<Node, std::io::Error> {
        let is_array = (prefix & 0x01) != 0;
        if is_array {
            self.decode_array(tag)
        } else {
            self.decode_object(tag)
        }
    }

    fn decode_array(&mut self, tag: &Tag) -> Result<Node, std::io::Error> {
        let l = self.read_byte()?;
        let len = l as usize;

        let mut items = Vec::new();
        for _ in 0..len {
            let item = self.decode_node(&Tag::new())?;
            items.push(item);
        }

        Ok(Node::Array(Array {
            items,
            tag: Some(Tag::new()),
        }))
    }

    fn decode_object(&mut self, tag: &Tag) -> Result<Node, std::io::Error> {
        let mut fields = Vec::new();

        let l = self.read_byte()?;
        let _len = l as usize;

        loop {
            if self.offset >= self.data.len() {
                break;
            }
            let b = self.data[self.offset];
            if b == 0x00 || b == 0x01 {
                break;
            }

            let key_node = self.decode_node(&Tag::new())?;
            if let Node::Value(v) = key_node {
                let key = v.text;
                let value = self.decode_node(&Tag::new())?;
                fields.push(Field { key, value });
            } else {
                break;
            }
        }

        Ok(Node::Object(Object {
            fields,
            tag: Some(Tag::new()),
        }))
    }
}

File: mm-kt/src/main/kotlin/io/github/metamessage/mm/Decoder.kt
package io.github.metamessage.mm

import java.math.BigInteger
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.UUID
import io.github.metamessage.ast.Tag
import io.github.metamessage.ast.ValueType
import io.github.metamessage.ast.Field
import io.github.metamessage.ast.Node
import io.github.metamessage.ast.Value
import io.github.metamessage.ast.Array
import io.github.metamessage.ast.Object

import io.github.metamessage.jsonc.toJsonc

class Decoder() {
    private var offset = 0
    private lateinit var data: ByteArray


    fun decode(data: ByteArray): Node {
        this.data = data
        this.offset = 0

        val d = decodeNode(null)
        if (offset != data.size) {
            throw MmDecodeException("trailing bytes at $offset len ${data.size}")
        }
        // println("22222222${toJsonc(d.node)}")
        return d.node
    }

    private data class Decoded(val node: Node, val consumed: Int)

    private fun decodeNode(inherited: Tag?): Decoded {
        val start = offset
        if (offset >= data.size) {
            throw MmDecodeException("eof")
        }
        val b = data[offset++].toInt() and 0xFF
        val prefix = b and Prefix.PREFIX_MASK
        val d = when (prefix) {
            Prefix.TAG -> decodeTagged(b, start)
            Prefix.SIMPLE -> Decoded(decodeSimple(b, inherited), offset - start)
            Prefix.POSITIVE_INT -> decodePositiveInt(b, inherited, start)
            Prefix.NEGATIVE_INT -> decodeNegativeInt(b, inherited, start)
            Prefix.FLOAT -> decodeFloat(b, inherited, start)
            Prefix.STRING -> decodeString(b, inherited, start)
            Prefix.BYTES -> decodeBytes(b, inherited, start)
            Prefix.CONTAINER -> decodeContainer(b, inherited, start)
            else -> throw MmDecodeException("invalid prefix")
        }
        return d
    }

    private fun decodeTagged(firstByte: Int, start: Int): Decoded {
        val tl = tagOuterLen(firstByte)
        val l1 = tl[0]
        var l2 = tl[1]
        if (l1 == 1) {
            l2 = data[offset++].toInt() and 0xFF
        } else if (l1 == 2) {
            l2 = ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
        val innerStart = offset
        val innerEnd = innerStart + l2
        if (innerEnd > data.size) {
            throw MmDecodeException("tag frame past eof")
        }
        val tb = data[offset++].toInt() and 0xFF
        val innerFieldLen = readInnerLen(tb)
        val fieldsEnd = offset + innerFieldLen
        if (fieldsEnd > innerEnd) {
            throw MmDecodeException("tag fields overflow")
        }
        val tag = Tag.empty()
        while (offset < fieldsEnd) {
            val n = TagFieldParser.parseOne(TagFieldParser.Cursor(data, offset), tag)
            if (n <= 0) {
                throw MmDecodeException("tag error")
            }
            offset += n
        }
        val node: Node
        if (tag.isNull) {
            val synthetic = nullScalarForTag(tag)
            if (synthetic != null) {
                node = synthetic
            } else {
                val inner = decodeNode(tag)
                node = inner.node
            }
        } else {
            val inner = decodeNode(tag)
            node = inner.node
        }
        // Ensure offset reaches innerEnd
        offset = innerEnd
        return Decoded(node, offset - start)
    }

    private fun nullScalarForTag(tag: Tag): Node? {
        return when (tag.type) {
            ValueType.DATETIME -> {
                val z = zoneForHours(tag.location)
                val dt = LocalDateTime.ofInstant(Instant.EPOCH, z ?: ZoneOffset.UTC)
                Value(dt, dt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")), tag)
            }
            ValueType.DATE -> {
                val d = LocalDate.of(1970, 1, 1)
                Value(d, d.toString(), tag)
            }
            ValueType.TIME -> Value(LocalTime.MIDNIGHT, LocalTime.MIDNIGHT.toString(), tag)
            ValueType.INT8 -> Value(0.toByte(), "0", tag)
            ValueType.INT16 -> Value(0.toShort(), "0", tag)
            ValueType.INT32 -> Value(0, "0", tag)
            ValueType.INT64 -> Value(0L, "0", tag)
            ValueType.UINT, ValueType.UINT8, ValueType.UINT16, ValueType.UINT32 -> Value(0, "0", tag)
            ValueType.UINT64 -> Value(0L, "0", tag)
            ValueType.FLOAT32 -> Value(0f, "0.0", tag)
            ValueType.FLOAT64 -> Value(0.0, "0.0", tag)
            ValueType.EMAIL, ValueType.UUID, ValueType.DECIMAL -> Value("", "", tag)
            ValueType.BIGINT -> Value(BigInteger.ZERO, "0", tag)
            ValueType.URL -> Value("", "", tag)
            ValueType.IP -> Value(null, ipNullText(tag.version), tag)
            else -> null
        }
    }

    private fun ipNullText(version: Int): String {
        return when (version) {
            4 -> "0.0.0.0"
            6 -> "::"
            else -> ""
        }
    }

    private fun zoneForHours(hours: Int): ZoneId? {
        return if (hours == 0) null else ZoneOffset.ofHours(hours)
    }

    private fun tagOuterLen(firstByte: Int): IntArray {
        val l = firstByte and WireConstants.TAG_LEN_MASK
        return when {
            l < WireConstants.TAG_LEN_1 -> intArrayOf(0, l)
            l == WireConstants.TAG_LEN_1 -> intArrayOf(1, 0)
            l == WireConstants.TAG_LEN_2 -> intArrayOf(2, 0)
            else -> intArrayOf(0, l)
        }
    }

    private fun readInnerLen(b: Int): Int {
        val l = b
        return when {
            l < 254 -> l
            l < 257 -> data[offset++].toInt() and 0xFF
            else -> ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
    }

    private fun decodeSimple(first: Int, inherited: Tag?): Node {
        val tag = inherited?.copy() ?: Tag.empty()
        val sv = first and Prefix.SUFFIX_MASK
        return when (sv) {
            SimpleValue.FALSE -> {
                tag.type = ValueType.BOOL
                Value(false, "false", tag)
            }
            SimpleValue.TRUE -> {
                tag.type = ValueType.BOOL
                Value(true, "true", tag)
            }
            SimpleValue.NULL_BOOL -> nullBool(tag)
            SimpleValue.NULL_INT -> nullInt(tag)
            SimpleValue.NULL_FLOAT -> nullFloat(tag)
            SimpleValue.NULL_STRING -> nullString(tag)
            SimpleValue.NULL_BYTES -> nullBytes(tag)
            else -> throw MmDecodeException("unsupported simple: $sv")
        }
    }

    private fun nullBool(tag: Tag): Node {
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.BOOL
        if (tag.type != ValueType.BOOL) throw MmDecodeException("null_bool type mismatch")
        return Value(false, "false", tag)
    }

    private fun nullInt(tag: Tag): Node {
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.INT
        if (tag.type != ValueType.INT) throw MmDecodeException("null_int type mismatch")
        return Value(0, "0", tag)
    }

    private fun nullFloat(tag: Tag): Node {
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.FLOAT64
        if (tag.type != ValueType.FLOAT32 && tag.type != ValueType.FLOAT64) {
            throw MmDecodeException("null_float type mismatch")
        }
        return if (tag.type == ValueType.FLOAT32) {
            Value(0f, "0.0", tag)
        } else {
            Value(0.0, "0.0", tag)
        }
    }

    private fun nullString(tag: Tag): Node {
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.STRING
        if (tag.type != ValueType.STRING) throw MmDecodeException("null_string type mismatch")
        return Value("", "", tag)
    }

    private fun nullBytes(tag: Tag): Node {
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.BYTES
        if (tag.type != ValueType.BYTES) throw MmDecodeException("null_bytes type mismatch")
        return Value(ByteArray(0), "", tag)
    }

    private fun decodePositiveInt(first: Int, inherited: Tag?, start: Int): Decoded {
        val v = readUintBody(first)
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.INT
        return Decoded(mapUintToTree(tag, v), offset - start)
    }

    private fun decodeNegativeInt(first: Int, inherited: Tag?, start: Int): Decoded {
        val v = readUintBody(first)
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.INT
        return Decoded(mapNegativeInt(tag, v), offset - start)
    }

    private fun readUintBody(first: Int): Long {
        val l1 = intLenExtraBytes(first)
        val low = first and WireConstants.INT_LEN_MASK
        val v = if (l1 == 0) {
            low.toLong()
        } else {
            var v = 0L
            for (i in 0 until l1) {
                v = (v shl 8) or (data[offset++].toInt() and 0xFF).toLong()
            }
            v
        }
        return v
    }

    private fun intLenExtraBytes(first: Int): Int {
        val l = first and WireConstants.INT_LEN_MASK
        return if (l < WireConstants.INT_LEN_1) 0 else l - WireConstants.INT_LEN_1 + 1
    }

    private fun mapUintToTree(tag: Tag, v: Long): Node {
        return when (tag.type) {
            ValueType.INT -> Value(v.toInt(), v.toString(), tag)
            ValueType.INT8 -> Value(v.toByte(), v.toString(), tag)
            ValueType.INT16 -> Value(v.toShort(), v.toString(), tag)
            ValueType.INT32 -> Value(v.toInt(), v.toString(), tag)
            ValueType.INT64 -> Value(v, v.toString(), tag)
            ValueType.UINT -> Value(v.toInt(), v.toString(), tag)
            ValueType.UINT8 -> Value(v.toShort(), v.toString(), tag)
            ValueType.UINT16 -> Value(v.toInt(), v.toString(), tag)
            ValueType.UINT32 -> Value(v.toInt(), v.toString(), tag)
            ValueType.UINT64 -> Value(v, v.toString(), tag)
            ValueType.DATETIME -> decodeDateTime(tag, v)
            ValueType.DATE -> decodeDate(tag, v)
            ValueType.TIME -> decodeTime(tag, v)
            ValueType.ENUM -> decodeEnum(tag, v)
            else -> throw MmDecodeException("unsupported int type: ${tag.type}")
        }
    }

    private fun mapNegativeInt(tag: Tag, v: Long): Node {
        return when (tag.type) {
            ValueType.INT -> Value((-v).toInt(), "-$v", tag)
            ValueType.INT8 -> Value((-v).toByte(), "-$v", tag)
            ValueType.INT16 -> Value((-v).toShort(), "-$v", tag)
            ValueType.INT32 -> Value((-v).toInt(), "-$v", tag)
            ValueType.INT64 -> Value(-v, "-$v", tag)
            else -> throw MmDecodeException("unsupported neg int type: ${tag.type}")
        }
    }

    private fun decodeDateTime(tag: Tag, v: Long): Node {
        if (tag.isNull) return Value(null, "", tag)
        val ins = Instant.ofEpochSecond(v)
        val z = zoneForHours(tag.location)
        val ldt = LocalDateTime.ofInstant(ins, z ?: ZoneOffset.UTC)
        return Value(ldt, ldt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")), tag)
    }

    private fun decodeDate(tag: Tag, v: Long): Node {
        if (tag.isNull) return Value(null, "", tag)
        if (v > Integer.MAX_VALUE) throw MmDecodeException("date overflow")
        val d = TimeUtil.dateFromDays(v)
        return Value(d, d.toString(), tag)
    }

    private fun decodeTime(tag: Tag, v: Long): Node {
        if (tag.isNull) return Value(null, "", tag)
        if (v > 86399) throw MmDecodeException("time out of range")
        val t = TimeUtil.timeFromSeconds(v.toInt())
        return Value(t, t.toString(), tag)
    }

    private fun decodeEnum(tag: Tag, v: Long): Node {
        if (tag.isNull) return Value(-1, "", tag)
        if (tag.enum.isEmpty()) throw MmDecodeException("enum without labels")
        val parts = tag.enum.split("\\|")
        if (v >= parts.size) throw MmDecodeException("enum index out of range")
        val label = parts[v.toInt()].trim()
        return Value(v.toInt(), label, tag)
    }

    private fun decodeFloat(first: Int, inherited: Tag?, start: Int): Decoded {
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.FLOAT64
        val l = first and WireConstants.FLOAT_LEN_MASK
        val `val`: Double = if (l < WireConstants.FLOAT_LEN_1) {
            var `val` = (first and 0xF).toDouble() / 10.0
            if ((first and WireConstants.FLOAT_NEG_MASK) != 0) `val` = -`val`
            `val`
        } else {
            val exp = data[offset++].toByte()
            val l1 = floatLenExtraBytes(first)
            val mantissa = if (l1 == 0) {
                0
            } else {
                var m = 0L
                for (i in 0 until l1) {
                    m = (m shl 8) or (data[offset++].toInt() and 0xFF).toLong()
                }
                m
            }
            val dec = FloatCodec.mantissaToDecimal(mantissa, exp)
            var `val` = dec.toDouble()
            if ((first and WireConstants.FLOAT_NEG_MASK) != 0) `val` = -`val`
            `val`
        }
        val node = when (tag.type) {
            ValueType.FLOAT32 -> Value(`val`.toFloat(), `val`.toString(), tag)
            ValueType.FLOAT64, ValueType.DECIMAL -> Value(`val`, `val`.toString(), tag)
            else -> throw MmDecodeException("bad float tag ${tag.type}")
        }
        return Decoded(node, offset - start)
    }

    private fun floatLenExtraBytes(first: Int): Int {
        val l = first and WireConstants.FLOAT_LEN_MASK
        return if (l < WireConstants.FLOAT_LEN_1) {
            0
        } else {
            l - WireConstants.FLOAT_LEN_1 + 1
        }
    }

    private fun decodeString(first: Int, inherited: Tag?, start: Int): Decoded {
        val sl = stringLen(first)
        var l2 = sl[1]
        if (sl[0] == 1) {
            l2 = data[offset++].toInt() and 0xFF
        } else if (sl[0] == 2) {
            l2 = ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
        val bs = if (l2 > 0) readBytes(l2) else ByteArray(0)
        val text = String(bs, StandardCharsets.UTF_8)
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.STRING
        val node = when (tag.type) {
            ValueType.STRING, ValueType.EMAIL -> Value(text, text, tag)
            ValueType.URL -> Value(text, text, tag)
            ValueType.IP -> Value(text, text, tag)
            else -> throw MmDecodeException("unsupported string type: ${tag.type}")
        }
        return Decoded(node, offset - start)
    }

    private fun stringLen(first: Int): IntArray {
        val l = first and WireConstants.STRING_LEN_MASK
        return when {
            l < WireConstants.STRING_LEN_1 -> intArrayOf(0, l)
            l == WireConstants.STRING_LEN_1 -> intArrayOf(1, 0)
            else -> intArrayOf(2, 0)
        }
    }

    private fun decodeBytes(first: Int, inherited: Tag?, start: Int): Decoded {
        val bl = bytesLen(first)
        var l2 = bl[1]
        if (bl[0] == 1) {
            l2 = data[offset++].toInt() and 0xFF
        } else if (bl[0] == 2) {
            l2 = ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
        val bs = if (l2 > 0) readBytes(l2) else ByteArray(0)
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.BYTES
        val node = when (tag.type) {
            ValueType.BYTES -> Value(bs, "", tag)
            ValueType.BIGINT -> bigintFromBytes(bs, tag)
            ValueType.UUID -> {
                if (bs.size != 16) throw MmDecodeException("uuid length")
                val u = uuidFromBytes(bs)
                Value(u, u.toString(), tag)
            }
            ValueType.IP -> Value(bs, "", tag)
            else -> throw MmDecodeException("unsupported bytes type: ${tag.type}")
        }
        return Decoded(node, offset - start)
    }

    private fun bytesLen(first: Int): IntArray {
        val l = first and WireConstants.BYTES_LEN_MASK
        return when {
            l < WireConstants.BYTES_LEN_1 -> intArrayOf(0, l)
            l == WireConstants.BYTES_LEN_1 -> intArrayOf(1, 0)
            else -> intArrayOf(2, 0)
        }
    }

    private fun uuidFromBytes(bs: ByteArray): UUID {
        val bb = ByteBuffer.wrap(bs)
        val hi = bb.long
        val lo = bb.long
        return UUID(hi, lo)
    }

    private fun bigintFromBytes(bs: ByteArray, tag: Tag): Node {
        if (bs.isEmpty()) return Value(BigInteger.ZERO, "0", tag)
        val n = bs[0].toInt() and 0xFF
        val body = bs.copyOfRange(1, bs.size)
        val bits = bigintBits(body)
        val neg = bits.isNotEmpty() && bits[0] == 1
        val digits = BigIntWireCodec.decodePositive(body, n)
        val bi = BigInteger(if (neg) "-$digits" else digits)
        return Value(bi, bi.toString(), tag)
    }

    private fun bigintBits(data: ByteArray): List<Int> {
        val bits = mutableListOf<Int>()
        for (bt in data) {
            for (i in 7 downTo 0) {
                bits.add((bt.toInt() shr i) and 1)
            }
        }
        return bits
    }

    private fun decodeContainer(first: Int, inherited: Tag?, start: Int): Decoded {
        val isArray = (first and WireConstants.CONTAINER_MASK) == WireConstants.CONTAINER_ARRAY
        val isMap = (first and WireConstants.CONTAINER_MASK) == WireConstants.CONTAINER_MAP
        return if (isArray) {
            decodeArray(first, inherited, start)
        } else if (isMap) {
            decodeObject(first, inherited, start)
        } else {
            decodeObject(first, inherited, start)
        }
    }

    private fun decodeArray(first: Int, inherited: Tag?, start: Int): Decoded {
        val cl = containerLen(first)
        var l2 = cl[1]
        if (cl[0] == 1) {
            l2 = data[offset++].toInt() and 0xFF
        } else if (cl[0] == 2) {
            l2 = ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
        val bodyStart = offset
        val bodyEnd = bodyStart + l2
        if (bodyEnd > data.size) {
            throw MmDecodeException("array past eof")
        }
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) {
            tag.type = if (tag.size > 0) ValueType.ARRAY else ValueType.SLICE
        }
        val items = mutableListOf<Node>()
        while (offset < bodyEnd) {
            val elemTag = Tag.empty()
            elemTag.inheritFromArrayParent(tag)
            val el = decodeNode(elemTag)
            items.add(el.node)
        }
        if (offset != bodyEnd) {
            throw MmDecodeException("array body misaligned")
        }
        return Decoded(Array(items, tag = tag), offset - start)
    }

    private fun decodeObject(first: Int, inherited: Tag?, start: Int): Decoded {
        val cl = containerLen(first)
        var l2 = cl[1]
        if (cl[0] == 1) {
            l2 = data[offset++].toInt() and 0xFF
        } else if (cl[0] == 2) {
            l2 = ((data[offset++].toInt() and 0xFF) shl 8) or (data[offset++].toInt() and 0xFF)
        }
        val innerStart = offset
        val innerEnd = innerStart + l2
        if (innerEnd > data.size) {
            throw MmDecodeException("object past eof")
        }
        val tag = inherited?.copy() ?: Tag.empty()
        if (tag.type == ValueType.UNKNOWN) tag.type = ValueType.STRUCT
        val keyPrefixPos = offset
        val keyPrefix = data[offset++].toInt() and 0xFF
        val keysDec = decodeArray(keyPrefix, null, keyPrefixPos)
        val keys = keysDec.node as Array
        val fields = mutableListOf<Field>()
        var i = 0
        while (offset < innerEnd && i < keys.items.size) {
            val elemTag = Tag.empty()
            elemTag.inheritFromArrayParent(tag)
            val `val` = decodeNode(elemTag)
            val key = (keys.items[i] as Value).text
            fields.add(Field(key, `val`.node))
            i++
        }
        // Ensure offset reaches innerEnd
        offset = innerEnd
        return Decoded(Object(fields, tag = tag), offset - start)
    }

    private fun containerLen(first: Int): IntArray {
        val l = first and WireConstants.CONTAINER_LEN_MASK
        return when {
            l < WireConstants.CONTAINER_LEN_1 -> intArrayOf(0, l)
            l == WireConstants.CONTAINER_LEN_1 -> intArrayOf(1, 0)
            else -> intArrayOf(2, 0)
        }
    }

    private fun readBytes(n: Int): ByteArray {
        if (offset + n > data.size) throw MmDecodeException("eof")
        val r = ByteArray(n)
        System.arraycopy(data, offset, r, 0, n)
        offset += n
        return r
    }
}
